### PR TITLE
Add type hints for mmcv/cnn/bricks

### PR DESCRIPTION
**Pull request by triple-Mu**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Add type hints for mmcv/cnn/bricks

## Modification

mmcv/cnn/bricks

## BC-breaking (Optional)

No

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

**Before PR**:

- [ ] I have read and followed the workflow indicated in the [CONTRIBUTING.md](https://github.com/open-mmlab/mmcv/blob/master/CONTRIBUTING.md) to create this PR.
- [ ] Pre-commit or linting tools indicated in [CONTRIBUTING.md](https://github.com/open-mmlab/mmcv/blob/master/CONTRIBUTING.md) are used to fix the potential lint issues.
- [ ] Bug fixes are covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] New functionalities are covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [ ] The documentation has been modified accordingly, including docstring or example tutorials.

**After PR**:

- [ ] If the modification has potential influence on downstream or other related projects, this PR should be tested with some of those projects, like MMDet or MMCls.
- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: mmcv/cnn/bricks/drop.py [hunk 1]
```diff
@@ -56,10 +60,10 @@ class Dropout(nn.Dropout):
         inplace (bool):  Do the operation inplace or not. Default: False.
     """
 
-    def __init__(self, drop_prob=0.5, inplace=False):
+    def __init__(self, drop_prob: float = 0.5, inplace: bool = False):
         super().__init__(p=drop_prob, inplace=inplace)
 
 
-def build_dropout(cfg, default_args=None):
+def build_dropout(cfg: Dict, default_args: Optional[Dict] = None):
```
Comment (HAOCHENYE) [inline mmcv/cnn/bricks/drop.py:67]: type hints for return value

File: mmcv/cnn/bricks/non_local.py [hunk 2]
```diff
@@ -224,9 +229,9 @@ class NonLocal1d(_NonLocalNd):
     """
 
     def __init__(self,
-                 in_channels,
-                 sub_sample=False,
-                 conv_cfg=dict(type='Conv1d'),
+                 in_channels: int,
+                 sub_sample: bool = False,
+                 conv_cfg: Optional[Dict] = dict(type='Conv1d'),
```
Comment (HAOCHENYE) [inline mmcv/cnn/bricks/non_local.py:234]: the default value of `conv_cfg` is not `None`

File: mmcv/cnn/bricks/non_local.py [hunk 3]
```diff
@@ -257,9 +262,9 @@ class NonLocal2d(_NonLocalNd):
     _abbr_ = 'nonlocal_block'
 
     def __init__(self,
-                 in_channels,
-                 sub_sample=False,
-                 conv_cfg=dict(type='Conv2d'),
+                 in_channels: int,
+                 sub_sample: bool = False,
+                 conv_cfg: Optional[Dict] = dict(type='Conv2d'),
```
Comment (HAOCHENYE) [inline mmcv/cnn/bricks/non_local.py:267]: same as above

File: mmcv/cnn/bricks/non_local.py [hunk 4]
```diff
@@ -287,9 +292,9 @@ class NonLocal3d(_NonLocalNd):
     """
 
     def __init__(self,
-                 in_channels,
-                 sub_sample=False,
-                 conv_cfg=dict(type='Conv3d'),
+                 in_channels: int,
+                 sub_sample: bool = False,
+                 conv_cfg: Optional[Dict] = dict(type='Conv3d'),
```
Comment (HAOCHENYE) [inline mmcv/cnn/bricks/non_local.py:297]: same as above

File: mmcv/cnn/bricks/padding.py [hunk 5]
```diff
@@ -8,7 +10,7 @@
 PADDING_LAYERS.register_module('replicate', module=nn.ReplicationPad2d)
 
 
-def build_padding_layer(cfg, *args, **kwargs):
+def build_padding_layer(cfg: Optional[Dict], *args, **kwargs) -> nn.Module:
```
Comment (HAOCHENYE) [inline mmcv/cnn/bricks/padding.py:13]: It seems the `cfg` must be a dict. Suggest to modify the docstring and type hints

File: mmcv/cnn/bricks/plugin.py [hunk 6]
```diff
@@ -48,12 +51,14 @@ def camel2snack(word):
...
             f'class_type must be a type, but got {type(class_type)}')
     if hasattr(class_type, '_abbr_'):
-        return class_type._abbr_
+        return class_type._abbr_  # type: ignore
     else:
         return camel2snack(class_type.__name__)
 
 
-def build_plugin_layer(cfg, postfix='', **kwargs):
+def build_plugin_layer(cfg: Optional[Dict],
```
Comment (HAOCHENYE) [inline mmcv/cnn/bricks/plugin.py:59]: same as above

File: mmcv/cnn/bricks/wrappers.py [hunk 7]
```diff
@@ -21,27 +21,28 @@
     TORCH_VERSION = tuple(int(x) for x in torch.__version__.split('.')[:2])
 
 
-def obsolete_torch_version(torch_version, version_threshold):
+def obsolete_torch_version(torch_version, version_threshold) -> bool:
```
Comment (HAOCHENYE) [inline mmcv/cnn/bricks/wrappers.py:24]: type hints

File: mmcv/image/colorspace.py [hunk 8]
```diff
@@ -272,11 +275,11 @@ def ycbcr2bgr(img):
     return out_img
 
 
-def convert_color_factory(src, dst):
+def convert_color_factory(src: str, dst: str) -> Callable:
 
     code = getattr(cv2, f'COLOR_{src.upper()}2{dst.upper()}')
 
-    def convert_color(img):
+    def convert_color(img: Union[np.ndarray, str]) -> np.ndarray:
```
Comment (HAOCHENYE) [inline mmcv/image/colorspace.py:282]: Why `img`  could be a string?

**Comment by zhouzaida** [comment]
Hi, the CI failed.

**Review by HAOCHENYE — APPROVED** [review]

File: mmcv/cnn/bricks/norm.py [hunk 9]
```diff
@@ -119,7 +122,8 @@ def build_norm_layer(cfg, num_features, postfix=''):
     return name, layer
 
 
-def is_norm(layer, exclude=None):
+def is_norm(layer: nn.Module,
+            exclude: Optional[Union[type, tuple]] = None) -> bool:
```
Comment (zhouzaida) [inline mmcv/cnn/bricks/norm.py:126]:
```suggestion
            exclude: Union[type, tuple, None] = None) -> bool:
```

File: mmcv/cnn/bricks/norm.py [hunk 10]
```diff
@@ -1,5 +1,6 @@
 # Copyright (c) OpenMMLab. All rights reserved.
 import inspect
+from typing import Dict, Optional, Tuple, Union
```
Comment (zhouzaida) [inline mmcv/cnn/bricks/norm.py:3]:
```suggestion
from typing import Dict, Tuple, Union
```

File: mmcv/cnn/bricks/wrappers.py [hunk 11]
```diff
@@ -21,27 +21,28 @@
...
     return torch_version == 'parrots' or torch_version <= version_threshold
 
 
 class NewEmptyTensorOp(torch.autograd.Function):
 
     @staticmethod
-    def forward(ctx, x, new_shape):
+    def forward(ctx: torch.Tensor, x: torch.Tensor,
+                new_shape: tuple) -> torch.Tensor:
```
Comment (zhouzaida) [inline mmcv/cnn/bricks/wrappers.py:32]:
```suggestion
    def forward(ctx: x: torch.Tensor, new_shape: tuple) -> torch.Tensor:
```

File: mmcv/cnn/bricks/wrappers.py [hunk 12]
```diff
@@ -21,27 +21,28 @@
...
         ctx.shape = x.shape
         return x.new_empty(new_shape)
 
     @staticmethod
-    def backward(ctx, grad):
+    def backward(ctx: torch.Tensor, grad: torch.Tensor) -> tuple:
```
Comment (zhouzaida) [inline mmcv/cnn/bricks/wrappers.py:37]:
```suggestion
    def backward(ctx, grad: torch.Tensor) -> tuple:
```

**Review by zhouzaida — APPROVED** [review]
